File: expectanim/src/main/java/com/github/florent37/expectanim/core/scale/ScaleAnimExpectation.java
package com.github.florent37.expectanim.core.scale;

import android.support.annotation.Nullable;
import android.view.View;

import com.github.florent37.expectanim.GravityScaleHorizontalIntDef;
import com.github.florent37.expectanim.GravityScaleVerticalIntDef;
import com.github.florent37.expectanim.core.AnimExpectation;
import com.github.florent37.expectanim.core.Expectations;
import com.github.florent37.expectanim.core.position.PositionAnimExpectation;

/**
 * Created by florentchampigny on 17/02/2017.
 */

public abstract class ScaleAnimExpectation extends AnimExpectation {

    protected boolean toDp = false;
    protected boolean keepRatio = false;
    @Nullable
    private Integer gravityHorizontal;
    @Nullable
    private Integer gravityVertical;

    public ScaleAnimExpectation(@Nullable Integer gravityHorizontal, @Nullable Integer gravityVertical) {
        if (gravityHorizontal != null) {
            this.gravityHorizontal = gravityHorizontal;
        }
        if (gravityVertical != null) {
            this.gravityVertical = gravityVertical;
        }
    }

    protected int dpToPx(float value, View view) {
        final int v =  (int) PositionAnimExpectation.dpToPx(view.getContext(), value);
        toDp = false;
        return v;
    }

    public abstract Float getCalculatedValueScaleX(View viewToMove);

    public abstract Float getCalculatedValueScaleY(View viewToMove);

    public Integer getGravityHorizontal() {
        return gravityHorizontal;
    }

    public Integer getGravityVertical() {
        return gravityVertical;
    }

    public ScaleAnimExpectation withGravity(@GravityScaleHorizontalIntDef @Nullable Integer gravityHorizontal, @GravityScaleVerticalIntDef @Nullable Integer gravityVertical) {
        if (gravityHorizontal != null) {
            this.gravityHorizontal = gravityHorizontal;
        }
        if (gravityVertical != null) {
            this.gravityVertical = gravityVertical;
        }
        return this;
    }

    public void setToDp(boolean toDp) {
        this.toDp = toDp;
    }

    public void setKeepRatio(boolean keepRatio) {
        this.keepRatio = keepRatio;
    }
}

File: expectanim/src/main/java/com/github/florent37/expectanim/core/scale/ScaleAnimExpectation.kt
package com.github.florent37.expectanim.core.scale

import android.view.View

import com.github.florent37.expectanim.GravityScaleHorizontalIntDef
import com.github.florent37.expectanim.GravityScaleVerticalIntDef
import com.github.florent37.expectanim.core.AnimExpectation
import com.github.florent37.expectanim.core.Expectations
import com.github.florent37.expectanim.core.position.PositionAnimExpectation

abstract class ScaleAnimExpectation(gravityHorizontal: Int?, gravityVertical: Int?) : AnimExpectation() {

    var toDp = false
    var keepRatio = false
    var gravityHorizontal: Int? = null
        private set
    var gravityVertical: Int? = null
        private set

    init {
        if (gravityHorizontal != null) {
            this.gravityHorizontal = gravityHorizontal
        }
        if (gravityVertical != null) {
            this.gravityVertical = gravityVertical
        }
    }

    protected fun dpToPx(value: Float, view: View): Int {
        val v = PositionAnimExpectation.dpToPx(view.context, value).toInt()
        toDp = false
        return v
    }

    abstract fun getCalculatedValueScaleX(viewToMove: View): Float?

    abstract fun getCalculatedValueScaleY(viewToMove: View): Float?

    fun withGravity(@GravityScaleHorizontalIntDef gravityHorizontal: Int?, @GravityScaleVerticalIntDef gravityVertical: Int?): ScaleAnimExpectation {
        if (gravityHorizontal != null) {
            this.gravityHorizontal = gravityHorizontal
        }
        if (gravityVertical != null) {
            this.gravityVertical = gravityVertical
        }
        return this
    }
}
